pkg/kubelet/kubelet_pods.go

```diff
@@ -144,11 +144,11 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 	// Kubernetes only mounts on /etc/hosts if:
 	// - container is not an infrastructure (pause) container
 	// - container is not already mounting on /etc/hosts
-	// - OS is not Windows
-	// - if it is Windows, ContainerD is used.
+	// - if it is Windows and ContainerD is used.
 	// Kubernetes will not mount /etc/hosts if:
 	// - when the Pod sandbox is being created, its IP is still unknown. Hence, PodIP will not have been set.
-	mountEtcHostsFile := len(podIPs) > 0 && supportsSingleFileMapping
+	// - Windows pod has a hostProcess container
+	mountEtcHostsFile := len(podIPs) > 0 && supportsSingleFileMapping && !kubecontainer.HasWindowsHostProcessContainer(pod)
```

Comment:
how does it validates the runtime used?

Reply:
`HasWindowsHostProcessContainer` is not checking the feature gate setting. Also it is not checking for Windows. So this check adds all containers iteration for linux as well. Perhaps small adjustment may be made to ensure that the existing logic is not affected in case feature gate is not set and it is not a windows

Reply:
`supportsSingleFileMapping` is calcutated here: https://github.com/kubernetes/kubernetes/blob/2e93b3924ae6eb2b6a2b914fe400a006bd676308/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L282-L285

Reply:
> in case feature gate is not set

this makes sense, I thought `podutil.AllFeatureEnabledContainers()` in https://github.com/kubernetes/kubernetes/blob/7705b300e2085c3864bb1e49a7302bf17f080219/pkg/kubelet/container/helpers.go#L315-L317

covered that but it looks like there are additional guards elsewhere https://github.com/kubernetes/kubernetes/blob/7705b300e2085c3864bb1e49a7302bf17f080219/pkg/kubelet/kuberuntime/labels.go#L109-L110

Reply:
I thought the same thing :(

Reply:
nit: Maybe we can comment it clearly for the better code maintenance?